File: module/app/src/main/java/es/ull/pcg/hpc/fancierfrontend_testapp/MainActivity.java
package es.ull.pcg.hpc.fancierfrontend_testapp;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.widget.ImageView;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import java.lang.reflect.Executable;
import java.util.concurrent.Executors;

import es.ull.pcg.hpc.fancierfrontend_testapp.examples.Example1_Basic;
import es.ull.pcg.hpc.fancierfrontend_testapp.examples.Example2_InPlace;
import es.ull.pcg.hpc.fancierfrontend_testapp.examples.Example3_Bitmap;
import es.ull.pcg.hpc.fancierfrontend_testapp.examples.Example4_FancierTypes;
import es.ull.pcg.hpc.fancierfrontend_testapp.examples.Example5_MultipleStages;
import timber.log.Timber;

public class MainActivity extends AppCompatActivity {

    private static ImageView imageViewBefore;
    private static ImageView imageViewAfter;
    private static MainActivity instance;

    public static void setImgBefore(Bitmap img) {
        instance.runOnUiThread(() -> MainActivity.imageViewBefore.setImageBitmap(img));
    }

    public static void setImgAfter(Bitmap img) {
        instance.runOnUiThread(() -> MainActivity.imageViewAfter.setImageBitmap(img));
    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (BuildConfig.DEBUG) {
            Timber.plant(new LinkingTree());
        }
        Context ctx = getApplicationContext();
        instance = this;
        setContentView(R.layout.activity_main);
        imageViewBefore = findViewById(R.id.imgBefore);
        imageViewAfter = findViewById(R.id.imgAfter);
        Executors.newSingleThreadExecutor().execute(() -> {
//            Example1_Basic.run(ctx);
//            Example2_InPlace.run(ctx);
//            Example3_Bitmap.run(ctx);
//            Example4_FancierTypes.run(ctx);
            Example5_MultipleStages.run(ctx);
        });
    }
}

File: examples/example_app/app/src/main/java/es/ull/pcg/hpc/fancyjcl_example/MainActivity.kt
package es.ull.pcg.hpc.fancyjcl_example

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import es.ull.pcg.hpc.fancyjcl_example.filters.*
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.newSingleThreadContext
import timber.log.Timber
import timber.log.Timber.Forest.plant

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (BuildConfig.DEBUG) {
            plant(LinkingTree())
        }
        ActivityCompat.requestPermissions(
            this, arrayOf(
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            ), 1
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 0) {
            if (grantResults.size != 1 || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                val alertDialog: android.app.AlertDialog? =
                    android.app.AlertDialog.Builder(this).create()
                alertDialog!!.setTitle("Camera Permission")
                alertDialog.setMessage("Camera Permission is not granted")
                alertDialog.setButton(
                    AlertDialog.BUTTON_NEUTRAL,
                    "OK"
                ) { dialog, _ -> dialog.dismiss() }
                alertDialog.show()
            }
        } else {
            initialize()
        }
    }

    private fun initialize() {
        setContentView(R.layout.activity_main)
        ctx = applicationContext
        MainActivity.layoutInflater =
            getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        instance = this
        scopeBackground = CoroutineScope(newSingleThreadContext("background_jobs"))
        scopeUI = CoroutineScope(Dispatchers.Main + SupervisorJob())
        val arrayList: ArrayList<Any> = ArrayList()
        arrayList.add(TabAdapter.Tests())
        arrayList.add(TabAdapter.Benchmarks())
        view_pager.adapter = TabAdapter(arrayList)

        TabLayoutMediator(
            tab_layout, view_pager
        ) { tab: TabLayout.Tab, position: Int ->
            tab.text = if (position == 0) "Tests" else "Benchmarks"
        }.attach()
    }

    companion object {
        lateinit var ctx: Context
        lateinit var layoutInflater: LayoutInflater
        lateinit var instance: MainActivity
        lateinit var scopeBackground: CoroutineScope
        lateinit var scopeUI: CoroutineScope
        fun getFilterByIndex(index: Int): Filter {
            return when (index) {
                0 -> Bilateral()
                1 -> Median()
                2 -> Posterize()
                3 -> Levels()
                4 -> Fisheye()
                5 -> Contrast()
                6 -> Convolution5x5()
                7 -> GrayScale()
                8 -> Convolution3x3()
                9 -> GaussianBlur()
                else -> {
                    throw Exception("No filter for index $index")
                }
            }
        }
    }
}